### PR TITLE
Calendar

### PR DESCRIPTION
**Pull request by anthonyqzhu**
Implemented code to display the users primary calendar in the cal.html page. This PR sets up the Oauth flows for the calendar API requests but does not deal with access tokens yet. The calendar ID is retrieved from the API and injected into the page via javascript.

This was tested by doing a test deployment of the app and navigating to the cal.html page after logging in

### CODE REVIEW

**Comment by anthonyqzhu** [comment]
Only the last 8 commits are new, I am not sure why older commits are included in the PR

File: src/main/webapp/script.js [hunk 1]
```diff
@@ -97,3 +97,19 @@ function deleteTask(task) {
...
+    fetch('/calendar').then((response) => response.text()).then((calSrc) => {
+        var calFrame = document.createElement('iframe');
+        calFrame.setAttribute('src', calSrc);
+        calFrame.setAttribute('style', 'border: 0'); 
+        calFrame.setAttribute('width', '800'); 
+        calFrame.setAttribute('height', '600'); 
+        calFrame.setAttribute('frameborder', '0'); 
+        calFrame.setAttribute('scrolling', 'no'); 
+        calElement.appendChild(calFrame)
+        calElement.appendChild(calFrame);
```
Comment (SeanT18) [inline src/main/webapp/script.js:172]: Are these supposed to repeat

Comment (anthonyqzhu) [reply src/main/webapp/script.js:172]: fixed in the next PR

File: src/main/java/com/google/sps/servlets/CalendarServlet.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  @Override
+  protected String getRedirectUri(HttpServletRequest req) throws ServletException, IOException {
+    return Utils.getRedirectUri(req);
+  }
+
+  @Override
+  protected AuthorizationCodeFlow initializeFlow() throws IOException {
+    return Utils.newFlow();
+  }
+}
```
Comment (SeanT18) [inline src/main/java/com/google/sps/servlets/CalendarServlet.java]: newline

File: src/main/java/com/google/sps/servlets/OAuth2Callback.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  @Override
+  protected String getRedirectUri(HttpServletRequest req) throws ServletException, IOException {
+    return Utils.getRedirectUri(req);
+  }
+
+  @Override
+  protected AuthorizationCodeFlow initializeFlow() throws IOException {
+    return Utils.newFlow();
+  }
+}
```
Comment (SeanT18) [inline src/main/java/com/google/sps/servlets/OAuth2Callback.java]: newline

File: src/main/java/com/google/sps/servlets/Utils.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  static IOException wrappedIOException(IOException e) {
+    if (e.getClass() == IOException.class) {
+      return e;
+    }
+    return new IOException(e.getMessage());
+  }
+
+  private Utils() {
+  }
+}
```
Comment (SeanT18) [inline src/main/java/com/google/sps/servlets/Utils.java]: newline

File: src/main/java/com/google/sps/servlets/Utils.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    return new Calendar.Builder(HTTP_TRANSPORT, JSON_FACTORY, credential).build();
+  }
+
+  /**
+   * Returns an {@link IOException} (but not a subclass) in order to work around restrictive GWT
+   * serialization policy.
+   */
+  static IOException wrappedIOException(IOException e) {
+    if (e.getClass() == IOException.class) {
+      return e;
```
Comment (SeanT18) [inline src/main/java/com/google/sps/servlets/Utils.java:95]: +2 indent

**Review by SeanT18 — APPROVED** [review]

File: src/main/java/com/google/sps/servlets/OAuth2Callback.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      throws ServletException, IOException {
+    resp.sendRedirect("/cal.html");
+  }
+
+  @Override
+  protected void onError(
+      HttpServletRequest req, HttpServletResponse resp, AuthorizationCodeResponseUrl errorResponse)
+      throws ServletException, IOException {
+    String nickname = UserServiceFactory.getUserService().getCurrentUser().getNickname();
+    resp.getWriter().print("<h3>" + nickname + ", why don't you want to play with me?</h1>");
```
Comment (yjennywang) [inline src/main/java/com/google/sps/servlets/OAuth2Callback.java:44]: Should this error message be more serious? :P

File: src/main/webapp/cal.html [hunk 7]
```diff
@@ -8,7 +8,8 @@
         <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Cinzel">
         <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lora">
     </head>
-    <body>
-      <iframe src="https://calendar.google.com/calendar/embed?src=aqzhu%40google.com" style="border: 0" width="800" height="600" frameborder="0" scrolling="no"></iframe>
+    <body onload="insertCalendar()">
+      <div id="caldiv">
```
Comment (yjennywang) [inline src/main/webapp/cal.html:12]: nit: no need to restate in the id that the div is a div

Just cal, maybe?

**Review by yjennywang — COMMENTED** [review]
Drive by comments after the fact.

**Pull request merged**